middle_end/flambda/from_lambda/cps_conversion.ml

```diff
@@ -325,76 +523,119 @@ let rec_catch_for_for_loop
                 Lstaticraise (cont, [next_value_of_counter]),
                 L.lambda_unit))),
           L.lambda_unit)))
-  in lam
+  in env, lam
 
-let rec cps_non_tail env (lam : L.lambda) (k : Ident.t -> Ilambda.t)
+let let_cont_with_extra_params env ~continuation_on_stack ~is_exn_handler
+      ~params ~(recursive : Asttypes.rec_flag)
+      ~(body : Env.t -> Continuation.t -> Ilambda.t)
+      ~(handler : Env.t -> Ilambda.t) : Ilambda.t =
+  let cont = Continuation.create () in
+  let { Env. body_env; handler_env; extra_params } =
+    Env.add_continuation env cont
+      ~push_to_try_stack:continuation_on_stack recursive
+  in
+  let extra_params =
+     List.map (fun (id, kind) -> id, Ilambda.User_visible, kind) extra_params
+  in
+  Let_cont {
+    name = cont;
+    is_exn_handler;
+    params = params @ extra_params;
+    recursive;
+    body = body body_env cont;
+    handler = handler handler_env;
+  }
+
+let apply_cont_with_extra_args env cont traps args =
+  let extra_args =
+      List.map (fun var : Ilambda.simple -> Var var)
+        (Env.extra_args_for_continuation env cont)
+  in
+  I.Apply_cont (cont, traps, args @ extra_args)
+
+let wrap_apply_with_extra_params env (apply : Ilambda.apply) =
```

Comment:
I think the name of this function is not ideal. It sounds like it's adding parameters to the function application itself, while it's only adding arguments to the return continuation. What about `wrap_return_continuation` ?